mne/decoding/ssd.py

```diff
@@ -0,0 +1,316 @@
+                             (n_chan, self.info['nchan']))
+
+    def fit(self, X, y=None):
+        """Estimate the SSD decomposition on raw or epoched data.
+
+        Parameters
+        ----------
+        X : array, shape ([n_epochs, ]n_channels, n_times)
+            The input data from which to estimate the SSD. Either 2D array
+            obtained from continuous data or 3D array obtained from epoched
+            data.
+        y : None | array, shape (n_samples,)
+            Used for scikit-learn compatibility.
+
+        Returns
+        -------
+        self : instance of SSD
+            Returns the modified instance.
+        """
+        self._check_X(X)
+        if X.ndim == 2:
+            X_aux = X[self.picks_, :].copy()
+
+        elif X.ndim == 3:
+            X_aux = X[:, self.picks_, :].copy()
+        else:
+            raise NotImplementedError()
+
+        X_signal = filter_data(
+            X_aux, self.info['sfreq'], **self.filt_params_signal)
+        X_noise = filter_data(
+            X_aux, self.info['sfreq'], **self.filt_params_noise)
+
+        X_noise -= X_signal
+        if X.ndim == 3:
+            X_signal = np.hstack(X_signal)
+            X_noise = np.hstack(X_noise)
+
+        cov_signal = _regularized_covariance(
+            X_signal, reg=self.reg, method_params=self.cov_method_params,
+            rank=self.rank, info=self.info)
+        cov_noise = _regularized_covariance(
+            X_noise, reg=self.reg, method_params=self.cov_method_params,
+            rank=self.rank, info=self.info)
+
+        eigvals_, eigvects_ = eigh(cov_signal, cov_noise)
+        # sort in descending order
+        ix = np.argsort(eigvals_)[::-1]
+        self.eigvals_ = eigvals_[ix]
+        self.filters_ = eigvects_[:, ix]
+        self.patterns_ = np.linalg.pinv(self.filters_)
+
+        return self
+
+    def transform(self, X, y=None):
+        """Estimate epochs sources given the SSD filters.
+
+        Parameters
+        ----------
+        X : array, shape ([n_epochs, ]n_channels, n_times)
+            The input data from which to estimate the SSD. Either 2D array
+            obtained from continuous data or 3D array obtained from epoched
+            data.
+        y : None | array, shape (n_samples,)
+            Used for scikit-learn compatibility.
+
+        Returns
+        -------
+        X_ssd : array, shape ([n_epochs, ]n_components, n_times)
+            The processed data.
+        """
+        self._check_X(X)
+        if self.filters_ is None:
+            raise RuntimeError('No filters available. Please first call fit')
+        if X.ndim == 2:
+            X_ssd = np.dot(self.filters_.T, X[self.picks_])
+        elif X.ndim == 3:
+            # project data on source space
+            X_ssd = np.empty_like(X)
+            for ii, x in enumerate(X[:, self.picks_]):
+                X_ssd[ii] = np.dot(self.filters_.T, x)
+        else:
+            raise NotImplementedError()
```

Comment:
Use broadcasting `@` operator
```suggestion
        X_ssd = self.filters_.T @ X[..., self.picks_, :]
```

Reply:
> Regarding coverage I suspect some of it is epochs vs raw. One way to fix this is to make the code path the same for both, which I've suggested. But even then it's good to test that it actually works in both cases. One way to do this is to do some parameterization of the test function with something like:
> 
> ```
> @pytest.mark.parametrize('kind', ('raw', 'epochs'))
> def test_whatever(kind):
>     n_channels, n_times = 10, 10000
>     info = create_info(n_channels, 1000., 'eeg')
>     data = np.random.RandomState(0).randn(n_channels, n_times)
>     if kind == 'raw':
>         inst = mne.io.RawArray(info, data)
>     else:
>         assert kind == 'epochs'
>         # use two replicates as singleton dims can errantly pass in ways
>         # that non-singletons wont
>         inst = mne.EpochsArray(info, np.repeat(data[np.newaxis], 2, axis=0))
>     ...
> ```

but does it not sufficient in the way we have it in test_ssd.py::test_ssd (line 106-)
```
filt_params_signal = dict(l_freq=freqs_sig[0], h_freq=freqs_sig[1],
                              l_trans_bandwidth=1, h_trans_bandwidth=1)
    filt_params_noise = dict(l_freq=freqs_noise[0], h_freq=freqs_noise[1],
                             l_trans_bandwidth=1, h_trans_bandwidth=1)
    ssd = SSD(info, filt_params_signal, filt_params_noise)
    raw = io.RawArray(X, info)

    pytest.raises(ValueError, ssd.fit, raw)
```
since we only supports np.array data, making the fit with a raw instance raises an error. Or is it not sufficient/similar?